http-server/README.md

```diff
@@ -71,3 +71,29 @@ wizard may fail to start with this error:
 java.lang.IllegalStateException: Unable to acquire the logger context
     at io.dropwizard.logging.LoggingUtil.getLoggerContext(LoggingUtil.java:46)
 ```
+
+## Stream is already closed
+
+This can happen when the server side does not fail fast on incorrect input.
+This can be if we use headers that are missing or do not check that parameters are present.
+
+The stack trace indicating this will look like this:
+```bash
+ERROR [2019-06-06 05:07:22,247] org.glassfish.jersey.server.ServerRuntime$Responder: An I/O error has occurred while writing a response message entity to the container output stream.
+! java.lang.IllegalStateException: The output stream has already been closed.
+```
+
+The impact of this is: 
+- server thread hangs
+- client hangs
+- server does not shutdown cleanly
+
+This is bad as it could be used in a DDOS attack.
+
+### Prevention
+
+Essentially fail-fast:
+- When looking for headers, verify headers exist or terminate the request
+- Verify that all needed GET parameters are present or terminate the request
+
+To terminate the request, just throw a IllegalArgumentException, it'l be mapped to a 400.
```

Comment:
This reminds me of a terrible habit in a project a have been working on in the past, where essentially everything (even NullPointerException/internal or external Server errors) got mapped to 400.
So I think getting a little bit more specific with the error code will not hurt. There are so many status codes to choose from for client errors, my favourite still being [418](https://httpstatuses.com/418).
For example [422](https://httpstatuses.com/422) is much more specific and indicates a semantic error whereas [400](https://httpstatuses.com/400) indicates a syntax error.

Reply:
>  (even NullPointerException/internal or external Server errors) got mapped to 400.

Ouch!

There should be some code that throws IllegalArg if an argument is missing. For the most part the syntax is not checked beyond arguments being present. Hence 422 is not appropriate. I would be relunctant to create our `ServerPrecondition.checkArg` method to throw a specific RuntimeException just for the purpose of mapping that to a 400.

One does need to be careful though about IllegalArgumentException. I've seen it abused where that exception is thrown due to programmer error at low levels in the service call stack. So essentially 2 or 3 layers deep, some service dependency fails, returns no data, that is then passed to another layer which throws illegal arg. That scenario should be a 500 as repeating the same exact request could eventually return a 200.

I did check around on 400s, does not seem that there are too many that are even more specific.  For now I think 400 and IllegalArg handling is okay, it was a good consideration but let's not also over-think it too much.